File: internal/integration/decision.go
package integration

import (
	"context"
	"errors"
	"fmt"
	"net"
)

// Outcome is the answer to a check.
type Outcome string

const (
	Allow   Outcome = "allow"
	Deny    Outcome = "deny"
	Unknown Outcome = "unknown"
)

// Code is a stable, machine-readable reason. The set is closed; integrations
// pick from it and never invent new ones.
type Code string

const (
	CodeAllowed            Code = "allowed"
	CodeDenied             Code = "denied"
	CodeUserNotFound       Code = "user_not_found"
	CodeUserAmbiguous      Code = "user_ambiguous"
	CodeUpstreamTimeout    Code = "upstream_timeout"
	CodeUpstreamError      Code = "upstream_error"
	CodeUpstreamRateLimit  Code = "upstream_rate_limited"
	CodeCredentialRejected Code = "credential_rejected"
	CodeResourceNotVisible Code = "resource_not_visible"
	CodeUnsupported        Code = "unsupported"
	CodeInvalidRequest     Code = "invalid_request"
	CodeUnknownConnection  Code = "unknown_connection"
	CodeUnknownAction      Code = "unknown_action"
	CodeUnauthorized       Code = "unauthorized"
)

// OutcomeOf maps a code to the only outcome it may carry. Allowed is allow,
// denied and user_not_found are deny, everything else is unknown.
func OutcomeOf(c Code) Outcome {
	switch c {
	case CodeAllowed:
		return Allow
	case CodeDenied, CodeUserNotFound:
		return Deny
	default:
		return Unknown
	}
}

// Decision is what a Connection returns from Check.
type Decision struct {
	Outcome Outcome
	Code    Code
	// Text is a short human-readable explanation. It must not contain secrets.
	Text string
	// Evidence is what the upstream system said when the decision was
	// computed. The engine fills it from the calls httpx recorded;
	// integrations leave it nil.
	Evidence *Evidence
}

// Reason renders "<code>: <text>", the wire form of the reason field.
func (d Decision) Reason() string {
	if d.Text == "" {
		return string(d.Code)
	}
	return string(d.Code) + ": " + d.Text
}

// Allowed builds an allow decision.
func Allowed(format string, args ...any) Decision {
	return Decision{Outcome: Allow, Code: CodeAllowed, Text: fmt.Sprintf(format, args...)}
}

// Denied builds a deny decision. The upstream system positively said no.
func Denied(format string, args ...any) Decision {
	return Decision{Outcome: Deny, Code: CodeDenied, Text: fmt.Sprintf(format, args...)}
}

// UnknownDecision builds an unknown decision with the given code.
// Codes whose outcome is not unknown are coerced to unsupported.
func UnknownDecision(code Code, format string, args ...any) Decision {
	if OutcomeOf(code) != Unknown {
		code = CodeUnsupported
	}
	return Decision{Outcome: Unknown, Code: code, Text: fmt.Sprintf(format, args...)}
}

// Unsupported is UnknownDecision with CodeUnsupported.
func Unsupported(format string, args ...any) Decision {
	return UnknownDecision(CodeUnsupported, format, args...)
}

// Error is an error that carries a decision code. Integrations return it from
// ResolveIdentity and Check when they cannot evaluate; the engine turns it
// into the matching decision. Wrapping keeps the cause for logs.
type Error struct {
	Code Code
	Text string
	Err  error
}

func (e *Error) Error() string {
	s := string(e.Code)
	if e.Text != "" {
		s += ": " + e.Text
	}
	if e.Err != nil {
		s += " (" + e.Err.Error() + ")"
	}
	return s
}

func (e *Error) Unwrap() error { return e.Err }

// Decision converts the error into a decision.
func (e *Error) Decision() Decision {
	return Decision{Outcome: OutcomeOf(e.Code), Code: e.Code, Text: e.Text}
}

// Errorf builds an *Error.
func Errorf(code Code, format string, args ...any) *Error {
	return &Error{Code: code, Text: fmt.Sprintf(format, args...)}
}

// Wrap builds an *Error around a cause.
func Wrap(code Code, err error, format string, args ...any) *Error {
	return &Error{Code: code, Text: fmt.Sprintf(format, args...), Err: err}
}

// UserNotFound is the error for an email with no account.
func UserNotFound(format string, args ...any) *Error {
	return Errorf(CodeUserNotFound, format, args...)
}

// UserAmbiguous is the error for an email that matches several accounts.
func UserAmbiguous(format string, args ...any) *Error {
	return Errorf(CodeUserAmbiguous, format, args...)
}

// ToDecision converts any error into a decision. *Error keeps its code;
// deadline and timeout errors become upstream_timeout; everything else is
// upstream_error. The wrapped cause is never put into the text.
func ToDecision(err error) Decision {
	var ie *Error
	if errors.As(err, &ie) {
		return ie.Decision()
	}
	if errors.Is(err, context.DeadlineExceeded) {
		return UnknownDecision(CodeUpstreamTimeout, "upstream call timed out")
	}
	var ne net.Error
	if errors.As(err, &ne) && ne.Timeout() {
		return UnknownDecision(CodeUpstreamTimeout, "upstream call timed out")
	}
	if errors.Is(err, context.Canceled) {
		return UnknownDecision(CodeUpstreamError, "request cancelled")
	}
	return UnknownDecision(CodeUpstreamError, "upstream call failed")
}

File: hallpass-py/src/hallpass/core/decision.py
"""The answer to a check: outcome, a closed set of reason codes, and errors
that carry a code.

The rule everything else rests on: only ``allowed`` is allow, ``denied``
and ``user_not_found`` are deny, and every other code is unknown. The
outcome is always derived from the code, never set on its own.
"""

from __future__ import annotations

import socket
from dataclasses import dataclass, field, replace
from enum import Enum
from typing import TYPE_CHECKING

from hallpass.core.context import Cancelled, DeadlineExceeded
from hallpass.core.errors import as_error, is_error

if TYPE_CHECKING:
    from hallpass.core.evidence import Evidence

__all__ = [
    "Code",
    "Decision",
    "HallpassError",
    "Outcome",
    "allowed",
    "denied",
    "errorf",
    "outcome_of",
    "to_decision",
    "unknown_decision",
    "unsupported",
    "user_ambiguous",
    "user_not_found",
    "wrap_error",
]


class Outcome(str, Enum):
    ALLOW = "allow"
    DENY = "deny"
    UNKNOWN = "unknown"

    def __str__(self) -> str:
        return self.value


class Code(str, Enum):
    """A stable, machine-readable reason. The set is closed."""

    ALLOWED = "allowed"
    DENIED = "denied"
    USER_NOT_FOUND = "user_not_found"
    USER_AMBIGUOUS = "user_ambiguous"
    UPSTREAM_TIMEOUT = "upstream_timeout"
    UPSTREAM_ERROR = "upstream_error"
    UPSTREAM_RATE_LIMIT = "upstream_rate_limited"
    CREDENTIAL_REJECTED = "credential_rejected"
    RESOURCE_NOT_VISIBLE = "resource_not_visible"
    UNSUPPORTED = "unsupported"
    INVALID_REQUEST = "invalid_request"
    UNKNOWN_CONNECTION = "unknown_connection"
    UNKNOWN_ACTION = "unknown_action"
    UNAUTHORIZED = "unauthorized"

    def __str__(self) -> str:
        return self.value


def outcome_of(code: Code | str | None) -> Outcome:
    """The only outcome a code may carry."""
    if code == Code.ALLOWED:
        return Outcome.ALLOW
    if code in (Code.DENIED, Code.USER_NOT_FOUND):
        return Outcome.DENY
    return Outcome.UNKNOWN


@dataclass(frozen=True)
class Decision:
    outcome: Outcome = Outcome.UNKNOWN
    code: Code | None = None
    # Text is a short human-readable explanation. It never contains
    # credential material or raw upstream bodies.
    text: str = ""
    # What the upstream system said when the decision was computed. The
    # engine fills it; integrations leave it None.
    evidence: Evidence | None = field(default=None, compare=False)

    def reason(self) -> str:
        """``"<code>: <text>"``, the wire form of the reason field."""
        code = "" if self.code is None else self.code.value
        if not self.text:
            return code
        return f"{code}: {self.text}"

    def with_(self, **changes: object) -> Decision:
        return replace(self, **changes)  # type: ignore[arg-type]


def allowed(text: str) -> Decision:
    return Decision(Outcome.ALLOW, Code.ALLOWED, text)


def denied(text: str) -> Decision:
    """The upstream system positively said no."""
    return Decision(Outcome.DENY, Code.DENIED, text)


def unknown_decision(code: Code, text: str) -> Decision:
    """An unknown decision. A code whose outcome is not unknown becomes unsupported."""
    if outcome_of(code) != Outcome.UNKNOWN:
        code = Code.UNSUPPORTED
    return Decision(Outcome.UNKNOWN, code, text)


def unsupported(text: str) -> Decision:
    return unknown_decision(Code.UNSUPPORTED, text)


class HallpassError(Exception):
    """An error that carries a decision code.

    Integrations raise it from resolve_identity and check when they cannot
    evaluate; the engine turns it into the matching decision. The cause
    (``__cause__``) is kept for logs and never put into the decision text.
    """

    def __init__(self, code: Code, text: str = "", cause: BaseException | None = None) -> None:
        self.code = code
        self.text = text
        self.cause = cause
        if cause is not None:
            self.__cause__ = cause
        super().__init__(str(self))

    def __str__(self) -> str:
        s = self.code.value
        if self.text:
            s += ": " + self.text
        if self.cause is not None:
            s += " (" + str(self.cause) + ")"
        return s

    def decision(self) -> Decision:
        return Decision(outcome_of(self.code), self.code, self.text)


def errorf(code: Code, text: str = "") -> HallpassError:
    return HallpassError(code, text)


def wrap_error(code: Code, err: BaseException | None, text: str = "") -> HallpassError:
    return HallpassError(code, text, err)


def user_not_found(text: str) -> HallpassError:
    """The error for an email with no account."""
    return HallpassError(Code.USER_NOT_FOUND, text)


def user_ambiguous(text: str) -> HallpassError:
    """The error for an email that matches several accounts."""
    return HallpassError(Code.USER_AMBIGUOUS, text)


def _is_timeout(err: BaseException) -> bool:
    for e in _walk(err):
        if isinstance(e, (socket.timeout, TimeoutError)) and not isinstance(e, DeadlineExceeded):
            return True
        t = getattr(e, "timeout", None)
        if callable(t):
            try:
                if t():
                    return True
            except Exception:  # noqa: BLE001
                pass
    return False


def _walk(err: BaseException):  # type: ignore[no-untyped-def]
    from hallpass.core.errors import chain

    return chain(err)


def to_decision(err: BaseException) -> Decision:
    """Any error as a decision.

    A HallpassError keeps its code; a spent deadline or a network timeout
    is upstream_timeout; everything else is upstream_error. The wrapped
    cause is never put into the text.
    """
    he = as_error(err, HallpassError)
    if he is not None:
        return he.decision()
    if is_error(err, DeadlineExceeded):
        return unknown_decision(Code.UPSTREAM_TIMEOUT, "upstream call timed out")
    if _is_timeout(err):
        return unknown_decision(Code.UPSTREAM_TIMEOUT, "upstream call timed out")
    if is_error(err, Cancelled):
        return unknown_decision(Code.UPSTREAM_ERROR, "request cancelled")
    return unknown_decision(Code.UPSTREAM_ERROR, "upstream call failed")
